src/fw/drivers/vibe/vibe_aw86225.c

```diff
@@ -156,8 +163,114 @@ int8_t vibe_get_braking_strength(void) {
+{
+  int brk_gain = 0x08;
+  int track_margain = 0x0F;
+  int f0 = 0;
+  int f0_pre = 235;
+  uint8_t drv_width = 0;
+  uint8_t reg_val = 0;
+
+  prv_write_register(0x5A, 0x00);
+  prv_modify_reg(0x08, ~(3 << 0), (2 << 0)); // Enter CONT mode
+  prv_modify_reg(0x18, ~(1 << 3), (1 << 3)); // Enable F0_DET
+  prv_modify_reg(0x1D, ~(1 << 7), (1 << 7)); // Enable track_en
+  prv_modify_reg(0x08, ~(1 << 2), (1 << 2)); // Enable auto brake
+  prv_modify_reg(0x1D, ~(0x7F << 0), 0x7F);  // Config cont_drv1_lvl
+  prv_write_register(0x1E, 0x7F); // Config cont_drv2_lvl
+  prv_write_register(0x1F, 0x04); // Config cont_drv1_time
+  prv_write_register(0x20, 0x06); // Config cont_drv2_time
+  prv_write_register(0x22, 0x0f); // Config cont_track_margin
+
+  drv_width = (uint8_t)(24000 / f0_pre - 8 - brk_gain - track_margain); // Calculate cont_drv_width
+  prv_write_register(0x1A, drv_width); // Config cont_drv_width
+
+  prv_write_register(0x09, 0x01); // Start play
+  psleep(300);
+  prv_write_register(0x09, 0x02); // Stop play
+
+  prv_read_register(0x25, &reg_val); // LRA_F0 high 8 bits
+  f0 = reg_val << 8;
+  prv_read_register(0x26, &reg_val); // LRA_F0 low 8 bits
+  f0 |= reg_val;
+  if (f0 == 0) {
+    return -1;
+  }
+  f0 = 384000 / f0;
+  
+  prv_modify_reg(0x18, ~(1 << 3), (0 << 3)); // Disable F0_DET
+  prv_modify_reg(0x08, ~(1 << 2), (0 << 2)); // Close auto brake
+
+  return f0;
+}
+
+static int vibe_f0_calibration(void) {
+  char f0_cali_lra = 0;
+  int f0_cali_step = 0;
+  int f0_pre = 235;
+  int f0_cali_min = 0;
+  int f0_cali_max = 0;
+  int f0 = 0;
+  
+  f0 = vibe_f0_detection();
+  if (f0 < 0) {
+    return -1;
+  }
+
+  /**
+   * TRIM_LRA[0:5] is used to calibration the f0 frequency.
+   *
+   * The TRIM_LRA LSB is 0.24%, up to 31, down to -32, calibration range
+   * is nearly +/- 7%.
+   *
+   * Below code calibrate the f0 to match f0_pre as possible.
+   */
+  f0_cali_min = f0_pre * 93 / 100;
+  f0_cali_max = f0_pre * 107 / 100;
+  if (f0 < f0_cali_min || f0 > f0_cali_max) {
+    return -1;
+  }
+
+  f0_cali_step = 10000 * ((int)f0 - (int)f0_pre) / ((int)f0_pre * 24);
+  if (f0_cali_step >= 0) {
+    if (f0_cali_step % 10 >= 5) {
+      f0_cali_step = 32 + (f0_cali_step / 10 + 1);
+    } else {
+      f0_cali_step = 32 + f0_cali_step / 10;
+    }
+
+  } else {
+    if (f0_cali_step % 10 <= -5) {
+      f0_cali_step = 32 + (f0_cali_step / 10 - 1);
+    } else {
+      f0_cali_step = 32 + f0_cali_step / 10;
+    }
+  }
+
+  if (f0_cali_step > 31) {
+    f0_cali_lra = (char)f0_cali_step - 32;
+  } else {
+    f0_cali_lra = (char)f0_cali_step + 32;
+  }
+
+  prv_write_register(0x5A, f0_cali_lra & 0x3F);
+
+  return 0;
+}
 
 void command_vibe_ctl(const char *arg) {
+  if (!strcmp(arg, "cal")) {
+    int rc = vibe_f0_calibration();
+    if (rc < 0) {
+      prompt_send_response("F0 cali fail");
+    } else {
+      prompt_send_response("F0 cali success");
+    }
+
+    return;
+  }
+
   int strength = atoi(arg);
```

Comment:
maybe add a new command, also, when/why should this be executed? maybe we should do it during initialization?